### PR TITLE
Password strength indication using bootstrap progress bar

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request is a redo for #8513
With some additions:
- Usage of JLayout
- Script version 2.0
- Removal of hardcoded staff

### CODE REVIEW

File: installation/form/field/password.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+JFormHelper::loadFieldClass('password');
+
+/**
+ * Form Field class for the Joomla Platform.
+ * Text field for passwords
+ *
+ * @link   http://www.w3.org/TR/html-markup/input.password.html#input.password
+ * @note   Two password fields may be validated as matching using JFormRuleEquals
+ * @since  11.1
```
Comment (zero-24) [inline installation/form/field/password.php]: since should be 3.6

File: installation/form/field/password.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ */
+class InstallationFormFieldPassword extends JFormFieldPassword
+{
+	/**
+	 * Method to get the field input markup for password.
+	 *
+	 * @return  string  The field input markup.
+	 *
+	 * @since   11.1
```
Comment (zero-24) [inline installation/form/field/password.php]: same here: since should be 3.6

File: libraries/joomla/form/fields/password.php [hunk 3]
```diff
@@ -44,6 +44,22 @@ class JFormFieldPassword extends JFormField
 	protected $maxLength;
 
 	/**
+	 * The allowable minlength of password.
+	 *
+	 * @var    integer
+	 * @since  3.5
```
Comment (zero-24) [inline libraries/joomla/form/fields/password.php]: since should be 3.6

**Comment by Kubik-Rubik** [comment]
Important for testers: Please test the installation process of Joomla! because we've encountered an issue during the Code Sprint in Manchester after we've merged the previous state of this PR (maybe it is already fixed?). Thanks!

File: installation/form/field/password.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	 */
+	protected function getInput()
+	{
+		$this->minLength = 4;
+		$return          = parent::getInput();
+
+		return $return;
```
Comment (zero-24) [inline installation/form/field/password.php]: maybe we can user here just: `return parent::getInput();`

File: installation/form/field/password.php (hunk 2 above)
Comment (dgt41) [inline installation/form/field/password.php]: Nope this is not a new function

File: libraries/joomla/form/fields/password.php (hunk 3 above)
Comment (dgt41) [inline libraries/joomla/form/fields/password.php]: yup

File: layouts/joomla/form/field/password.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+			};
+			jQuery('#" . $id . "').pwstrength(options);
+		});
+		"
+	);
+}
+?>
+<input type="password" name="<?php echo $name; ?>" id="<?php echo $id; ?>" value="<?php
+echo htmlspecialchars($value, ENT_COMPAT, 'UTF-8'); ?>" <?php
+echo $hint, $autocomplete, $class, $readonly, $disabled, $size, $maxLength,  $required, $autofocus; ?> />
```
Comment (zero-24) [inline layouts/joomla/form/field/password.php]: is there a reasone to have too much spaces bevor $required? ` $maxLength,  $required`

**Comment by zero-24** [comment]
@dgt41 does the script also support min integers, min symbols and min uppercase too?

File: installation/form/field/password.php (hunk 2 above)
Comment (zero-24) [inline installation/form/field/password.php]: you have add this file with 3.6 ;) this is what the `@since` tag indicates :smile: or i missunderstand?

**Comment by zero-24** [comment]
### How to test

##### First test
- Install that patch
- go to the user manger in backend
- add a new user
- see the Password strength indication for the password field.
- try to enter the username as password
- see the message
- set the minimum lenth of the password to 20
- try a password
- see the message

##### seccond test

- Install this as new branch https://github.com/dgt41/joomla-cms/archive/installationPass.zip
- see the Password strength indication for the password field.
- make sure you can install it like normal.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10377">issues.joomla.org/joomla-cms/10377</a>.</sub>

**Comment by dgt41** [comment]
@zero-24 these are the default rules
```javascript
defaultOptions.rules.activated = {
     wordNotEmail: true,
     wordLength: true,
     wordSimilarToUsername: true,
     wordSequences: true,
     wordTwoCharacterClasses: false,
     wordRepetitions: false,
     wordLowercase: true,
     wordUppercase: true,
     wordOneNumber: true,
     wordThreeNumbers: true,
     wordOneSpecialChar: true,
     wordTwoSpecialChar: true,
     wordUpperLowerCombo: true,
     wordLetterNumberCombo: true,
     wordLetterNumberCharCombo: true
 };
```
So there are a lot of options/combinations but the rules are specific to 1,2 or 3 uppercase,number,special character so overriding these we can justify the rule for 1,2 or 3 e.g. special characters. 
Thanks for the test instructions

**Comment by brianteeman** [comment]
What are character classes?
> Use different character classes

**Comment by brianteeman** [comment]
Will not install. Installer not loading correctly no header etc in the html

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 7f900873278c2458fd6fc291bdb646d5f0b67f03<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10377">issues.joomla.org/joomla-cms/10377</a>.</sub>

**Comment by 810** [comment]
@zero-24 with you package it is failing:
warning
JLIB_APPLICATION_ERROR_COMPONENT_NOT_LOADING

**Comment by zero-24** [comment]
@801 what do you meam? It is just the repo of this PR?!

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 4fad4f291f340c57380fe1d0a8968842b1f8d68d<br /><br />Go to installation - complete step 1 and click next to go to step 2
Decide you want to change the password so click on the first tab to return to step 1
Password script is no longer in use on this page
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10377">issues.joomla.org/joomla-cms/10377</a>.</sub>

**Comment by brianteeman** [comment]
Also if your confirm password is not match the password the javascript is no longer loaded<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10377">issues.joomla.org/joomla-cms/10377</a>.</sub>

**Comment by 810** [comment]
@zero-24 srry I thought you forked his pr.

@test I can still use admin - admin

**Comment by brianteeman** [comment]
If when registering a user you get a message such as
> Your password cannot contain your username

Then I would expect not to be able to save the user with that password
Currently you can which kind of defeats the object of this to me<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10377">issues.joomla.org/joomla-cms/10377</a>.</sub>

**Comment by brianteeman** [comment]
This is not ready for testing at all. Not even sure if it usable in its
current form

On 10 May 2016 at 13:54, Jelle Kok <notifications@github.com> wrote:

> @zero-24 <https://github.com/zero-24> srry I thought you forked his pr.
>
> @test <https://github.com/test> I can still use admin - admin
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/10377#issuecomment-218148394>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by dgt41** [comment]
@brianteeman was the previous behaviour like the one you are describing above or was it just an indication? If not I guess you are right and we should match the old behaviour

File: libraries/joomla/form/fields/password.php [hunk 6]
```diff
@@ -160,7 +160,7 @@ public function setup(SimpleXMLElement $element, $value, $group = null)
 
 			$this->minLength = $this->element['minlength'] ? (int) $this->element['minlength'] : $defaultMinLength;
 			$this->maxLength = $this->element['maxlength'] ? (int) $this->element['maxlength'] : 99;
-			$this->threshold = $this->element['threshold'] ? (int) $this->element['threshold'] : 66;
+			$this->threshold = $this->element['threshold'] ? (int) $this->element['threshold'] : 0;
```
Comment (dgt41) [inline libraries/joomla/form/fields/password.php]: Threshold is negative if rules don't match. The old value 66 was against different code

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on dfdee29ba3b7483f7a14a5eec00fa8a12e020cb0<br /><br />Problems I reported above about the installation are still not fixed



Sorry I wont be wasting my time testing this again. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10377">issues.joomla.org/joomla-cms/10377</a>.</sub>

**Comment by roland-d** [comment]
Hello @dgt41 

Thank you for your contribution.

The last comment here was on May 10th. Can you update this PR please?
If no reply is received within 4 weeks we will close this issue.

Thanks for understanding! <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10377">issues.joomla.org/joomla-cms/10377</a>.</sub>

**Comment by dgt41** [comment]
I will redo this with own script to match the exact old behaviour. Thanks

**Pull request closed without merging**